### PR TITLE
feat: implement generic functions

### PR DESCRIPTION
**Pull request by Shanks0224**

### CODE REVIEW

**Comment by Shanks0224** [comment]
PTAL @xujuntwt95329

File: src/scope.ts [hunk 1]
```diff
@@ -521,6 +531,7 @@ export class Scope {
         scope.localIndex = this.localIndex;
         scope.mangledName = this.mangledName;
         scope.modifiers = this.modifiers;
+        if (this.genericOwner) scope.setGenericOwner(this.genericOwner);
```
Comment (xujuntwt95329) [inline src/scope.ts:534]: `copy` and `clone`, which is used for generic specialization?
According to the function name, I would expect the result scope is the same as the original one (e.g. the genericOwner field should be the same), but this line breaks this assumption.
I would suggest to use `specialize` as the method name used to process generic specialization.

File: src/scope.ts [hunk 2]
```diff
@@ -671,12 +683,10 @@ export class FunctionScope extends ClosureEnvironment {
     /* ori func name iff func is declare */
     oriFuncName: string | undefined = undefined;
     debugLocations: SourceMapLoc[] = [];
-    // iff this FunctionScope is specialized
-    private _genericOwner?: FunctionScope;
 
-    constructor(parent: Scope) {
+    constructor(parent: Scope | null) {
```
Comment (xujuntwt95329) [inline src/scope.ts:687]: Why would it be possible to have a null parent scope?

File: src/semantics/expression_builder.ts [hunk 3]
```diff
@@ -921,6 +923,14 @@ function buildArrayLiteralExpression(
 
     // return new NewLiteralArrayValue(array_type!, init_values);
     const elem_type = (array_type as ArrayType).element;
+    // Workaround: solve the case that the return value of a generic function is a generic array
```
Comment (xujuntwt95329) [inline src/semantics/expression_builder.ts:926]: Why this only work for `return value of a generic function`? Seems it should also work for assignment statement?

File: src/semantics/expression_builder.ts [hunk 4]
```diff
@@ -1276,6 +1286,14 @@ export function newCastValue(
         return value;
     }
 
+    // Workaround: solve the case that the return value of a generic function is a generic array
```
Comment (xujuntwt95329) [inline src/semantics/expression_builder.ts:1289]: Same as above

File: src/semantics/index.ts [hunk 5]
```diff
@@ -205,8 +206,27 @@ function createFunctionDeclareNode(
...
+                v.varModifiers,
+                v.varIndex,
+                v.isOptional,
+                v.destructuring,
+            );
+            if (v.initExpression) newParam.setInitExpr(v.initExpression);
+            newParam.setIsLocalVar(v.isLocalVar());
+            newParam.needReBinding = v.needReBinding;
+            newParam.tsNode = v.tsNode;
+            f.addParameter(newParam);
```
Comment (yviansu) [inline src/semantics/index.ts:229]: So the specialized function scope is created in frontend, and the parameters are added in semantic tree, and it is same as class specialization too?

File: src/utils.ts [hunk 6]
```diff
@@ -427,17 +427,20 @@ export function processEscape(str: string) {
  * @param originalClassScope the original ClassScope to be specialized
  * @param parent the parent of the original ClassScope
  * @param classType the new class type corresponding to specialized ClassScope: TSClass => ClassScope
+ * @param newName the name of new ClassScope
  * @returns a new specialized ClassScope
  */
 export function createClassScopeByClassType(
     originalClassScope: ClassScope,
-    parent: Scope,
+    parent: Scope | null,
```
Comment (yviansu) [inline src/utils.ts:435]: In which situation, `parent` will be `null`?

File: src/scope.ts (hunk 1 above)
Comment (Shanks0224) [reply src/scope.ts:534]: done.
The name of method was changed to **specialize**

File: src/scope.ts (hunk 2 above)
Comment (Shanks0224) [reply src/scope.ts]: done.
This does seem redundant and wrong

File: src/semantics/expression_builder.ts (hunk 3 above)
Comment (Shanks0224) [reply src/semantics/expression_builder.ts]: done.
In **buildArrayLiteralExpression**, I made common logic to process generic arrays.It does not only work for return value of a generic function.

File: src/semantics/expression_builder.ts (hunk 4 above)
Comment (Shanks0224) [reply src/semantics/expression_builder.ts]: done.
This code has been deleted

File: src/semantics/index.ts (hunk 5 above)
Comment (Shanks0224) [reply src/semantics/index.ts:229]: Yes，the same goes for generic member functions in generic class.

File: src/utils.ts (hunk 6 above)
Comment (Shanks0224) [reply src/utils.ts]: done.
This does seem redundant and wrong.

**Review by yviansu — APPROVED** [review]
LGTM

**Review by xujuntwt95329 — APPROVED** [review]
LGTM

**Comment by xujuntwt95329** [comment]
@Shanks0224 Thanks for the great work!

**Pull request merged**